File: code/data_structures/src/CircularLinkedList/src_java/circularlinkedlist.java
class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class CircularLinkedList {
    private Node head;

    // Constructor to create an empty circular linked list
    CircularLinkedList() {
        head = null;
    }

    // Function to insert a node at the end of the circular linked list
    void append(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            head.next = head; // Point back to itself for a single node circular list
        } else {
            Node temp = head;
            while (temp.next != head) {
                temp = temp.next;
            }
            temp.next = newNode;
            newNode.next = head;
        }
    }

    // Function to delete a node by value from the circular linked list
    void delete(int data) {
        if (head == null)
            return;

        // If the head node contains the value to be deleted
        if (head.data == data) {
            Node temp = head;
            while (temp.next != head)
                temp = temp.next;
            if (head == head.next) {
                head = null; // If there is only one node
            } else {
                head = head.next;
                temp.next = head;
            }
            return;
        }

        // Search for the node to delete
        Node current = head;
        Node prev = null;
        while (current.next != head && current.data != data) {
            prev = current;
            current = current.next;
        }

        // If the node with the given data was found
        if (current.data == data) {
            prev.next = current.next;
        } else {
            System.out.println("Node with data " + data + " not found in the list.");
        }
    }

    // Function to display the circular linked list
    void display() {
        if (head == null)
            return;

        Node temp = head;
        do {
            System.out.print(temp.data + " ");
            temp = temp.next;
        } while (temp != head);
        System.out.println();
    }

    public static void main(String[] args) {
        CircularLinkedList cll = new CircularLinkedList();
        cll.append(1);
        cll.append(2);
        cll.append(3);

        System.out.print("Circular Linked List: ");
        cll.display(); // Output: 1 2 3

        cll.delete(2);
        System.out.print("After deleting 2: ");
        cll.display(); // Output: 1 3
    }
}

File: code/data_structures/src/CircularLinkedList/circularLinkedList.cpp
#include <iostream>
using namespace std;

/**
 * A node will be consist of a int data &
 * a reference to the next node.
 * The node object will be used in making linked list
 */
class Node
{
public:
    int data;
    Node *next;
};

/**
 * Using nodes for creating a circular linked list.
 *
 */
class CircularList
{
public:
    /**
     * stores the reference to the last node
     */
    Node *last;

    /**
     * keeping predecessor reference while searching
     */

    Node *preLoc;

    /**
     * current node reference while searching
     */
    Node *loc;

    CircularList()
    {
        last = NULL;
        loc = NULL;
        preLoc = NULL;
    }

    /**
     * Checking if list empty
     */
    bool isEmpty()
    {
        return last == NULL;
    }

    /**
     * Inserting new node in front
     */
    void insertAtFront(int value)
    {
        Node *newnode = new Node();
        newnode->data = value;

        if (isEmpty())
        {
            newnode->next = newnode;
            last = newnode;
        }
        else
        {
            newnode->next = last->next;
            last->next = newnode;
        }
    }

    /**
     * Inserting new node in last
     */
    void insertAtLast(int value)
    {
        Node *newnode = new Node();
        newnode->data = value;

        if (isEmpty())
        {
            newnode->next = newnode;
            last = newnode;
        }
        else
        {
            newnode->next = last->next;
            last->next = newnode;
            last = newnode;
        }
    }

    /**
     * Printing whole list iteratively
     */
    void printList()
    {
        if (!isEmpty())
        {
            Node *temp = last->next;
            do
            {
                cout << temp->data;
                if (temp != last)
                    cout << " -> ";
                temp = temp->next;
            } while (temp != last->next);
            cout << endl;
        }
        else
            cout << "List is empty" << endl;
    }

    /**
     * Searching a value
     */
    void search(int value)
    {
        loc = NULL;
        preLoc = NULL;
        if (isEmpty())
            return;
        loc = last->next;
        preLoc = last;
        while (loc != last && loc->data < value)
        {
            preLoc = loc;
            loc = loc->next;
        }
        if (loc->data != value)
        {
            loc = NULL;
            // if(value > loc->data)
            //     preLoc = last;
        }
    }

    /**
     * Inserting the value in its sorted postion in the list.
     */
    void insertSorted(int value)
    {
        Node *newnode = new Node();
        newnode->data = value;

        search(value);

        if (loc != NULL)
        {
            cout << "Value already exist!" << endl;
            return;
        }
        else
        {
            if (isEmpty())
                insertAtFront(value);
            else if (value > last->data)
                insertAtLast(value);
            else if (value < last->next->data)
                insertAtFront(value);
            else
            {
                newnode->next = preLoc->next;
                preLoc->next = newnode;
            }
        }
    }

    void deleteValue(int value)
    {
        search(value);
        if (loc != NULL)
        {
            if (loc->next == loc)
            {
                last = NULL;
            }
            else if (value == last->data)
            {
                preLoc->next = last->next;
                last = preLoc;
            }
            else
            {
                preLoc->next = loc->next;
            }
            delete loc;
        }
        else
            cout << "Value not found" << endl;
    }

    void destroyList()
    {
        Node *temp = last->next;
        while (last->next != last)
        {
            temp = last->next;
            last->next = last->next->next;
            delete temp;
        }
        delete last; // delete the only remaining node
        last = NULL;
    }
};

/**
 * Driver Code
 */
int main()
{

    CircularList *list1 = new CircularList();

    cout << "Initializing the list" << endl;
    list1->insertAtLast(2);
    list1->insertAtLast(3);
    list1->insertAtLast(4);
    list1->insertAtLast(6);
    list1->insertAtLast(8);

    list1->printList();

    cout << "Inserting 5 in the list sorted" << endl;
    list1->insertSorted(5);
    list1->printList();

    cout << "Deleting 3 from the list" << endl;
    list1->deleteValue(3);
    list1->printList();

    cout << "Destroying the whole list" << endl;
    list1->destroyList();
    list1->printList();
}